evap/contributor/views.py

```diff
@@ -21,22 +23,42 @@ def index(request):
-    own_evaluations = Evaluation.objects.filter(contributions__contributor=user, state__in=contributor_visible_states)
+    own_evaluations = Evaluation.objects.filter(
+        Q(state__in=contributor_visible_states) & (
+            Q(course__responsibles__in=[user]) |
+            Q(course__evaluations__contributions__contributor=user)
+        )
+    )
+    for evaluation in own_evaluations:
+        evaluation.contributes_to = evaluation.contributions.filter(contributor=user).exists()
 
-    displayed_evaluations = list(own_evaluations)
+    displayed_evaluations = set(own_evaluations)
     if show_delegated:
         represented_users = user.represented_users.all()
-        delegated_evaluations = Evaluation.objects.exclude(id__in=own_evaluations).filter(contributions__can_edit=True, contributions__contributor__in=represented_users, state__in=contributor_visible_states)
+        delegated_evaluations = Evaluation.objects.exclude(id__in=own_evaluations).filter(
+            Q(state__in=contributor_visible_states) & (
+                Q(course__responsibles__in=represented_users) |
+                Q(course__evaluations__contributions__can_edit=True, course__evaluations__contributions__contributor__in=represented_users)
+            )
+        )
         for evaluation in delegated_evaluations:
             evaluation.delegated_evaluation = True
-        displayed_evaluations += list(delegated_evaluations)
-    displayed_evaluations.sort(key=lambda evaluation: list(STATES_ORDERED.keys()).index(evaluation.state))
-
-    delegate_selection_form = DelegateSelectionForm()
+        displayed_evaluations.update(delegated_evaluations)
+    displayed_evaluations = list(displayed_evaluations)
+    displayed_evaluations.sort(key=lambda evaluation: evaluation.full_name)  # evaluations must be sorted for regrouping them in the template
 
     for evaluation in displayed_evaluations:
-        evaluation.distribution = calculate_average_distribution(evaluation)
-        evaluation.avg_grade = distribution_to_grade(evaluation.distribution)
+        if not evaluation.is_single_result:
+            evaluation.distribution = calculate_average_distribution(evaluation)
+            evaluation.avg_grade = distribution_to_grade(evaluation.distribution)
+        else:
+            evaluation.single_result_rating_result = get_single_result_rating_result(evaluation)
+        if evaluation.course.evaluations.count() > 1:
+            if any(evaluation.state != "published" for evaluation in evaluation.course.evaluations.all()):
+                evaluation.course.not_all_evaluations_are_published = True
+            evaluation.course.distribution = calculate_average_course_distribution(evaluation.course)
+            evaluation.course.avg_grade = distribution_to_grade(evaluation.course.distribution)
+            evaluation.course_weight_sum = sum(evaluation.weight for evaluation in evaluation.course.evaluations.all())
```

Comment:
We're having some kind of independent logic to determine what can be seen by this contributor and to calculate grades here. I think we should rather try to rely on the functions we already have that test whether people can see stuff or not - otherwise, semantics may drift apart. What do you think? (This might be something for another issue.)

Reply:
i will change it for this pr. we should have a (single) clear way of determining the visibility, otherwise it will be very hard to maintain. so i'll think about a more centralized way.

Reply:
please check again. grade calculation is now done in the results tools for all views.
collecting the evaluations which can be seen depends on the view (contributor index, student index, results index) so probably we'll keep having that in three places. what do you think?